air/src/execution_step/instructions/fold/utils.rs

```diff
@@ -104,10 +106,21 @@ pub(crate) fn create_canon_stream_map_iterable_value(
         return Ok(FoldIterableScalar::Empty);
     }
 
-    // TODO: this one is a relatively long operation and will be refactored in Boxed Value
+    // TODO: this one is a relatively heavy operation and will be refactored in Boxed Value
     // Can not create iterable from existing CanonStreamMap b/c CSM contains a map with
     // a limited lifetime but the boxed value needs static lifetime.
-    let values = canon_stream_map.iter().cloned().collect::<Vec<_>>();
+    let mut met_keys = HashSet::new();
+    let mut values = vec![];
+
+    for val in canon_stream_map.canon_stream_map.iter().rev() {
+        if let Some(map_key) = StreamMapKey::from_kvpair_owned(val) {
+            if !met_keys.contains(&map_key) {
```

Comment:
It would be slightly more performant with the entry API (no double hashing and lookup), but I'm not sure it worth it.